### PR TITLE
[risk=low][RW-6688] Prevent notebook copy between tiers in the UI

### PR DESCRIPTION
**Pull request by jmthibault79**
Description:

UI counterpart to #4908 (but not dependent on it)

Update the Copy Modal logic to prevent copies across access tiers.  Keep the existing logic of allowing Notebook copies across CDR Versions within an access tier and preventing Concept Set copies across CDR Versions.

copy modal tiers

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java [hunk 1]
```diff
@@ -284,6 +284,7 @@ private WorkspaceResource buildRecentResource(
 
   private void buildFromDbWorkspace(WorkspaceResource resource, DbWorkspace dbWorkspace) {
     resource.setCdrVersionId(commonMappers.cdrVersionToId(dbWorkspace.getCdrVersion()));
+    resource.setAccessTierShortName(dbWorkspace.getCdrVersion().getAccessTier().getShortName());
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java:287]: just above this: // TODO RW-5638 reimplement all this in MapStruct

hard agree

File: api/src/main/java/org/pmiops/workbench/utils/mappers/WorkspaceMapper.java [hunk 2]
```diff
@@ -141,6 +141,7 @@ void mergeResearchPurposeIntoWorkspace(
   @Mapping(target = "workspaceFirecloudName", source = "dbWorkspace.firecloudName")
   @Mapping(target = "workspaceBillingStatus", source = "dbWorkspace.billingStatus")
   @Mapping(target = "cdrVersionId", source = "dbWorkspace.cdrVersion")
+  @Mapping(target = "accessTierShortName", source = "dbWorkspace.cdrVersion.accessTier.shortName")
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/utils/mappers/WorkspaceMapper.java:144]: see?  look at how nice this is, and the compiler enforces it.

File: ui/src/app/components/copy-modal.spec.tsx [hunk 3]
```diff
@@ -13,6 +13,32 @@ import {WorkspacesApiStub} from 'testing/stubs/workspaces-api-stub';
 import {cdrVersionTiersResponse, CdrVersionsStubVariables} from 'testing/stubs/cdr-versions-api-stub';
 
 import {CopyModalComponent, CopyModalProps, CopyModalState} from './copy-modal';
+import {AccessTierShortNames} from 'app/utils/access-tiers';
+
+function simulateSelect(wrapper: ReactWrapper, reactSelect: Select, selection: string) {
```
Comment (jmthibault79) [inline ui/src/app/components/copy-modal.spec.tsx:18]: same idea as react-test-helpers/simulateSelection() but incompatible.  TODO!

File: ui/src/app/components/copy-modal.spec.tsx [hunk 4]
```diff
@@ -184,20 +223,11 @@ describe('CopyModal', () => {
...
-    // Select an option
-    wrapper.find(Select).find({type: 'option'})
-        .findWhere(e => e.text() === altCdrWorkspace.name)
-        .first()
-        .simulate('click');
+    simulateWorkspaceSelection(wrapper, altCdrWorkspace);
 
     expect(getNotebookCdrMismatchWarning(wrapper).getDOMNode().textContent).toBe(
         'The selected destination workspace uses a different dataset version ' +
-        `(${CdrVersionsStubVariables.ALT_WORKSPACE_CDR_VERSION}) than the current workspace ` +
```
Comment (jmthibault79) [inline ui/src/app/components/copy-modal.spec.tsx:200]: a passing-by text change

File: ui/src/app/components/copy-modal.tsx [hunk 5]
```diff
@@ -112,14 +117,20 @@ const NotebookCdrMismatch = (props: {text: string}) =>
       </FlexRow>
     </div>;
 
-const ConceptSetRestrictionText = () => <div style={styles.conceptSetsRestriction}>
+const ConceptSetRestrictionText = () => <div style={styles.restriction}>
   Concept sets can only be copied to workspaces using the same CDR version.
```
Comment (jmthibault79) [inline ui/src/app/components/copy-modal.tsx:121]: may want to add "access tier" here too.  It's implicit in this statement but that may not be obvious to users.

File: ui/src/app/components/copy-modal.tsx [hunk 6]
```diff
@@ -289,41 +299,52 @@ class CopyModalComponent extends React.Component<Props, State> {
...
+    const accessTierMismatch: boolean = fromAccessTierShortName !== destination.accessTierShortName;
+    const isNotebook: boolean = resourceType === ResourceType.NOTEBOOK;
+    const isConceptSet: boolean = resourceType === ResourceType.CONCEPTSET;
 
-    if (resourceType === ResourceType.NOTEBOOK) {
-      this.setNotebookCdrMismatchWarning(destination, fromCdrVersionId);
-    } else if (resourceType === ResourceType.CONCEPTSET) {
-      this.setConceptSetCdrMismatchError(destination, fromCdrVersionId);
-    }
+    cond([!accessTierMismatch && !cdrVersionMismatch, () => this.clearMismatchErrors(destination)],
```
Comment (jmthibault79) [inline ui/src/app/components/copy-modal.tsx:338]: I like what `cond()` lets me do here

File: ui/src/app/components/copy-modal.tsx [hunk 7]
```diff
@@ -289,41 +299,52 @@ class CopyModalComponent extends React.Component<Props, State> {
...
 
+  clearMismatchErrors(destination: Workspace) {
+    this.setState({accessTierMismatch: '', cdrMismatch: '', destination: destination});
+  }
+
   // OK to copy a notebook with a mismatch, but show a warning message
   setNotebookCdrMismatchWarning(destination: Workspace, fromCdrVersionId: string) {
     const warningMsg = `The selected destination workspace uses a different dataset version ` +
-        `(${this.cdrName(destination.cdrVersionId)}) than the current workspace (${this.cdrName(fromCdrVersionId)}). ` +
+        `(${this.cdrName(destination.cdrVersionId)}) from the current workspace (${this.cdrName(fromCdrVersionId)}). ` +
```
Comment (jmthibault79) [inline ui/src/app/components/copy-modal.tsx:309]: opportunistic grammar update

File: api/src/test/java/org/pmiops/workbench/api/UserMetricsControllerTest.java [hunk 8]
```diff
@@ -66,17 +70,16 @@
   @Autowired private CohortMapper cohortMapper;
   @Autowired private CommonMappers commonMappers;
   @Autowired private FirecloudMapper firecloudMapper;
+  @Autowired private AccessTierDao accessTierDao;
+  @Autowired private CdrVersionDao cdrVersionDao;
 
-  private FakeClock fakeClock = new FakeClock(NOW);
+  private final FakeClock fakeClock = new FakeClock(NOW);
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/UserMetricsControllerTest.java:76]: some opportunistic cleanup in this file, some of which was requested by codacy

File: ui/src/app/components/copy-modal.tsx [hunk 9]
```diff
@@ -333,16 +354,16 @@ class CopyModalComponent extends React.Component<Props, State> {
           onChange={(destWorkspace) => this.validateAndSetDestination(destWorkspace)}
           isOptionDisabled={(option) => this.isDestinationSameWorkspace(option.value)}
         />
-        {cdrMismatch && resourceType === ResourceType.CONCEPTSET && <ConceptSetCdrMismatch text={cdrMismatch}/>}
+        {accessTierMismatch && <AccessTierMismatch text={accessTierMismatch}/>}
+        {!accessTierMismatch && cdrMismatch && resourceType === ResourceType.CONCEPTSET && <ConceptSetCdrMismatch text={cdrMismatch}/>}
```
Comment (NehaBroad) [inline ui/src/app/components/copy-modal.tsx:358]: nit: this (line number 358 and 366) condition can be moved to a function, which can accept a ResourceType, eg
function (rt) {
return !accessTierMismatch && cdrMismatch && resourceType === rt;
}

Comment (jmthibault79) [reply ui/src/app/components/copy-modal.tsx]: sgtm

**Review by NehaBroad — APPROVED** [review]
👍 Tested:  Notebook & Concept Set (different CDR and Access Tier)

**Pull request merged**